src/hb-buffer.cc

```diff
@@ -1755,6 +1755,22 @@ hb_buffer_append (hb_buffer_t *buffer,
   memcpy (buffer->info + orig_len, source->info + start, (end - start) * sizeof (buffer->info[0]));
   if (buffer->have_positions)
     memcpy (buffer->pos + orig_len, source->pos + start, (end - start) * sizeof (buffer->pos[0]));
+
+  if (source->content_type == HB_BUFFER_CONTENT_TYPE_UNICODE)
+  {
+    if (start == 0 && orig_len == 0 && source->context_len[0] && !buffer->context_len[0])
```

Comment:
If `start > 0 && orig_len == 0` then part of `source` before `start` should also be added as `pre_context`, followed by `source`'s `pre_context`.

Reply:
Shouldn’t it be `source`’s pre-context followed by the part of `source` before `start`?

Reply:
Depends on which direction you look at. Pre-context moves backward, so I said it in the order it would be coded.